allennlp/modules/token_embedders/pretrained_transformer_embedder.py

```diff
@@ -1,41 +1,191 @@
+            tensor = F.pad(tensor, [0, length_to_pad], value=0)
+            # Shape: [batch_size * num_segments, self._max_length]
+            return tensor.reshape(-1, self._max_length)
+
+        return fold(token_ids), fold(segment_concat_mask)
+
+    def _unfold_long_sequences(
+        self, embeddings: torch.FloatTensor, batch_size: int, num_segment_concat_wordpieces: int
+    ) -> torch.FloatTensor:
+        """
+        We take 2D segments of a long sequence and flatten them out to get the whole sequence
+        representation while remove unnecessary special tokens.
+
+        [ [ [CLS]_emb A_emb B_emb C_emb [SEP]_emb ], [ [CLS]_emb D_emb E_emb [SEP]_emb [PAD]_emb ] ]
+        -> [ [CLS]_emb A_emb B_emb C_emb D_emb E_emb [SEP]_emb ]
+
+        We truncate the start and end tokens for all segments, recombine the segments,
+        and manually add back the start tokens. We generally don't need to add back
+        the end tokens -- because the last segment is usually shorter than self._max_length,
+        the final end tokens usually won't be touched in our truncation process.
+
+        There are special cases where sequences have a full last segment. The only issue
+        in this case is that the [SEP] embedding will be wrong. Nevertheless, [SEP] embeddings
+        are rarely, if ever, used, so it shouldn't be a huge problem. However, we do remedy
+        this case when such sequences with a full last segment are the longest ones in a batch
+        by adding back the last token representations.
+
+        # Parameters
+
+        embeddings: `torch.FloatTensor`
+            Shape: [batch_size * num_segments, self._max_length, embedding_size].
+        batch_size: `int`
+        num_segment_concat_wordpieces: `int`
+            The length of the original "[ [CLS] A B C [SEP] [CLS] D E F [SEP] ]", i.e.
+            the original `token_ids.size(1)`.
+
+        # Returns:
+
+        embeddings: `torch.FloatTensor`
+            Shape: [batch_size, self._num_wordpieces, embedding_size].
+        """
+        num_segments = int(embeddings.size(0) / batch_size)
+        embedding_size = embeddings.size(2)
+
+        # We want to remove all segment-level special tokens but maintain sequence-level ones
+        num_wordpieces = num_segment_concat_wordpieces - (num_segments - 1) * self._num_added_tokens
+
+        embeddings = embeddings.reshape(batch_size, num_segments * self._max_length, embedding_size)
+        # Shape: (batch_size, self._num_added_start_tokens, embedding_size)
+        start_token_embeddings = embeddings[:, : self._num_added_start_tokens, :]
+        # See comment above -- remedy when the longest sequences have full last segments.
+        # This is not necessarily the end token embeddings when sequences aren't full.
+        # Shape: (batch_size, self._num_added_end_tokens, embedding_size)
+        last_token_embeddings = embeddings[:, -self._num_added_end_tokens :, :]
```

Comment:
This is not correct. You need to use something like https://github.com/allenai/allennlp/blob/master/allennlp/nn/util.py#L1207 where the end indices can be computed from the mask.